Add guidance on packaging, releasing and versioning

milanmlft opened the pull request:
## Addresses #48

Some general guidance on packaging, releasing and versioning.

### Still missing

- Deployment

milanmlft committed:
Outline topics

milanmlft committed:
Add containerisation guide

milanmlft committed:
Add release management guide

milanmlft committed:
Add versioning guide

nikloynes commented on coding_standards/c++/README.md:
```diff
@@ -30,6 +30,18 @@ general.
+- **Source distribution**: Tag releases on GitHub, users build from source
+- **Package managers**: Submit to [vcpkg](https://vcpkg.io/), [Conan](https://conan.io/), or
+  [conda-forge](https://conda-forge.org/)
+- **System packages**: Create `.deb`, `.rpm`, or other system-specific packages
```
what about non-linux? might make sense to elaborate on this a little.

nikloynes reviewed (approved):
looks great! some minor comments to consider, but not particularly important.

milanmlft committed:
Add Sam's suggestion to use `@deprecated` decorator

Co-authored-by: Sam Cunliffe <samcunliffe@users.noreply.github.com>

samcunliffe commented on project_organisation/versioning/README.md:
```diff
@@ -0,0 +1,189 @@
+
+Specify version constraints for your dependencies to ensure compatibility:
+
+- **Python** (`requirements.txt` or `pyproject.toml`):
```
V. Good point. Edited above comment.

milanmlft committed:
Better suggestions for python dependency-tracking

Co-authored-by: Sam Cunliffe <samcunliffe@users.noreply.github.com>

samcunliffe reviewed (changes requested):
A bit of a faux-pas since you didn't invite me to review in the first place.

Buuuuut,  I think amalgamating and `rm`-ing the docs in `arc-internal-documentation` needed for this PR (the `rm` can come later, but amalgamation should happen before merge imo).

milanmlft committed:
Link to Python tooling for packaging tools

milanmlft committed:
Amalgamate recommendations from
https://github.com/UCL-ARC/arc-internal-documentation/blob/main/arc-teams/Collaborations/processes/projects/group-practices.md#coding-conventions

milanmlft committed:
Move language-specific stuff to the right place

samcunliffe commented on coding_standards/python/README.md:
```diff
@@ -16,6 +16,7 @@ standard. We also encourage compliance with modern style recommendations, such a
-for recommended tools.
+Python code should be released as ["wheels"](https://packaging.python.org/en/latest/discussions/wheel-vs-egg/) using
+[`build`](https://pypa-build.readthedocs.io/en/stable/index.html), and, ideally, be published through the
+[Python Package Index](https://pypi.python.org/pypi) or [conda-forge](https://conda-forge.org/). We maintain a
```
👍

samcunliffe commented on project_organisation/release/README.md:
```diff
@@ -181,6 +181,9 @@ For widely-used research software, consider maintaining LTS versions:
 - Clearly communicate which versions receive support
 - Define an end-of-life policy
 
+In the version control system, maintain release branches for publicly released versions of the code, so that bug fixes
+can be efficiently backported to earlier releases.
+
```
Interesting! I used to do this, but lately I "just" tag main. Because `Flat is better than nested`?
Anyway, leave as is the recommendation is probably better.

milanmlft replied:
Tbh I do the same, but this was in the ARC internal docs, and I could see it being useful if you have some long-term support for older versions

samcunliffe commented on project_organisation/release/README.md:
```diff
@@ -0,0 +1,227 @@
+
+## Citation
+
+An appropriate [CITATION](https://www.software.ac.uk/blog/encouraging-citation-software-introducing-citation-files) file
```
```suggestion
An appropriate [CITATION.cff](https://www.software.ac.uk/blog/encouraging-citation-software-introducing-citation-files) file
```

Marginal comment, but just to be explicit.

samcunliffe reviewed (approved):
👍 🫐 

I'd vote to merge this as-is. It's been open for 5 months, so everyone else has had plenty of time to review.

milanmlft committed:
Explicit citation file extension

Co-authored-by: Sam Cunliffe <samcunliffe@users.noreply.github.com>

milanmlft committed:
Fix links

milanmlft committed:
Does anyone even use JORS?

milanmlft committed:
202 is fine, lychee
